File: Java/01_Basics/05_Collections.java
// ─────────────────────────────────────────
//  COLLECTIONS  (ArrayList, HashMap, HashSet, Queue, Stack)
// ─────────────────────────────────────────

import java.util.*;
import java.util.stream.*;

public class Collections {
    public static void main(String[] args) {

        // ════════════════════════════════════
        //  ArrayList<T>  — dynamic array, ordered, allows duplicates
        // ════════════════════════════════════
        List<String> fruits = new ArrayList<>(List.of("apple", "banana", "cherry"));

        // Add
        fruits.add("mango");
        fruits.add(1, "avocado");            // insert at index
        fruits.addAll(List.of("kiwi", "plum"));

        // Access
        System.out.println(fruits.get(0));   // apple
        System.out.println(fruits.size());   // 7
        System.out.println(fruits.get(fruits.size() - 1));  // last element

        // Update
        fruits.set(0, "APPLE");             // replace at index

        // Remove
        fruits.remove("mango");             // by value
        fruits.remove(0);                   // by index

        // Search
        System.out.println(fruits.contains("banana"));  // true
        System.out.println(fruits.indexOf("banana"));   // index or -1

        // Sort
        Collections.sort(fruits);           // ascending
        fruits.sort(Comparator.reverseOrder());         // descending
        fruits.sort(Comparator.comparingInt(String::length)); // by length

        // Iterate
        for (String fruit : fruits)
            System.out.print(fruit + " ");
        System.out.println();

        fruits.forEach(f -> System.out.print(f + " "));
        System.out.println();

        // ════════════════════════════════════
        //  HashMap<K,V>  — key-value pairs, no guaranteed order
        // ════════════════════════════════════
        Map<String, Integer> scores = new HashMap<>();
        scores.put("Alice", 90);
        scores.put("Bob", 85);
        scores.put("Carol", 92);

        // Access
        System.out.println(scores.get("Alice"));         // 90
        System.out.println(scores.get("Dave"));          // null — no exception
        System.out.println(scores.getOrDefault("Dave", 0)); // 0

        // Add / Update
        scores.put("Dave", 78);            // add
        scores.put("Alice", 95);           // update
        scores.putIfAbsent("Bob", 99);     // only if key doesn't exist

        // Remove
        scores.remove("Dave");

        // Check
        System.out.println(scores.containsKey("Bob"));     // true
        System.out.println(scores.containsValue(95));      // true

        // Iterate
        for (Map.Entry<String, Integer> entry : scores.entrySet())
            System.out.println(entry.getKey() + ": " + entry.getValue());

        scores.forEach((k, v) -> System.out.println(k + " → " + v));

        System.out.println(scores.keySet());    // Set of keys
        System.out.println(scores.values());    // Collection of values

        // LinkedHashMap — preserves insertion order
        Map<String, Integer> ordered = new LinkedHashMap<>();
        ordered.put("z", 1); ordered.put("a", 2); ordered.put("m", 3);
        System.out.println(ordered);  // {z=1, a=2, m=3} — insertion order kept

        // TreeMap — sorted by key
        Map<String, Integer> sorted = new TreeMap<>(scores);
        System.out.println(sorted);   // alphabetical order

        // ════════════════════════════════════
        //  HashSet<T>  — unique values, no order
        // ════════════════════════════════════
        Set<Integer> setA = new HashSet<>(Set.of(1, 2, 3, 4, 5));
        Set<Integer> setB = new HashSet<>(Set.of(4, 5, 6, 7, 8));

        setA.add(6);
        setA.remove(1);
        System.out.println(setA.contains(3));   // true

        // Set operations (need copies — these modify in place)
        Set<Integer> union = new HashSet<>(setA);
        union.addAll(setB);
        System.out.println(union);

        Set<Integer> intersection = new HashSet<>(setA);
        intersection.retainAll(setB);
        System.out.println(intersection);

        Set<Integer> diff = new HashSet<>(setA);
        diff.removeAll(setB);
        System.out.println(diff);

        // LinkedHashSet — preserves insertion order
        Set<String> linked = new LinkedHashSet<>(List.of("banana", "apple", "cherry"));
        System.out.println(linked);   // [banana, apple, cherry] — order kept

        // TreeSet — sorted
        Set<String> tree = new TreeSet<>(linked);
        System.out.println(tree);    // [apple, banana, cherry]

        // ════════════════════════════════════
        //  Queue<T>  — FIFO (first in, first out)
        // ════════════════════════════════════
        Queue<String> queue = new LinkedList<>();
        queue.offer("first");
        queue.offer("second");
        queue.offer("third");

        System.out.println(queue.peek());    // first — doesn't remove
        System.out.println(queue.poll());    // first — removes it
        System.out.println(queue.size());    // 2

        // ════════════════════════════════════
        //  Deque  — double-ended queue (Stack + Queue)
        // ════════════════════════════════════
        Deque<Integer> stack = new ArrayDeque<>();
        stack.push(1);   // push to front
        stack.push(2);
        stack.push(3);

        System.out.println(stack.peek());    // 3 — top of stack
        System.out.println(stack.pop());     // 3 — removes from front
        System.out.println(stack.size());    // 2

        // ════════════════════════════════════
        //  Unmodifiable collections (Java 9+)
        // ════════════════════════════════════
        List<String> immutableList = List.of("a", "b", "c");
        Map<String, Integer> immutableMap = Map.of("x", 1, "y", 2);
        Set<Integer> immutableSet = Set.of(1, 2, 3);
        // immutableList.add("d");  // UnsupportedOperationException
    }
}

File: CPP/01_Basics/05_Collections.cpp
// ─────────────────────────────────────────
//  COLLECTIONS  (map, set, queue, stack, unordered variants)
// ─────────────────────────────────────────

#include <iostream>
#include <map>
#include <unordered_map>
#include <set>
#include <unordered_set>
#include <queue>
#include <stack>
#include <list>
#include <deque>
using namespace std;

int main() {

    // ════════════════════════════════════
    //  map<K,V>  — key-value, sorted by key, unique keys
    // ════════════════════════════════════
    map<string, int> scores;
    scores["Alice"] = 90;
    scores["Bob"]   = 85;
    scores["Carol"] = 92;
    scores.insert({"Dave", 78});
    scores.emplace("Eve", 88);    // more efficient insert

    // Access
    cout << scores["Alice"] << endl;     // 90
    // scores["Unknown"] creates a new key with value 0 — be careful!

    // Safe access
    auto it = scores.find("Bob");
    if (it != scores.end())
        cout << it->first << ": " << it->second << endl;

    cout << scores.count("Alice") << endl;    // 1 (exists)
    cout << scores.count("Zara")  << endl;    // 0 (not found)

    // Update
    scores["Alice"] = 95;

    // Remove
    scores.erase("Dave");

    // Iterate — sorted by key alphabetically
    for (auto& [key, val] : scores)   // structured bindings (C++17)
        cout << key << ": " << val << endl;
    cout << endl;

    cout << "Size: " << scores.size() << endl;

    // ════════════════════════════════════
    //  unordered_map<K,V>  — hash map, O(1) average, no order
    // ════════════════════════════════════
    unordered_map<string, int> freq;
    string text = "hello world hello cpp hello";
    for (char c : text)
        if (c != ' ') freq[string(1, c)]++;

    // Count word frequency
    unordered_map<string, int> wordCount;
    wordCount["apple"]++;
    wordCount["banana"]++;
    wordCount["apple"]++;
    wordCount["apple"]++;

    for (auto& [word, count] : wordCount)
        cout << word << ": " << count << endl;

    // ════════════════════════════════════
    //  set<T>  — unique sorted values
    // ════════════════════════════════════
    set<int> s = {5, 3, 1, 4, 2, 3, 5};  // duplicates removed on insert
    cout << "Set: ";
    for (int n : s) cout << n << " ";    // 1 2 3 4 5 (sorted)
    cout << endl;

    s.insert(6);
    s.insert(3);           // duplicate — ignored
    s.erase(1);
    cout << s.count(4) << endl;   // 1 (exists)
    cout << s.count(99)<< endl;   // 0

    auto sit = s.find(4);
    if (sit != s.end())
        cout << "Found: " << *sit << endl;

    // lower_bound / upper_bound
    auto lb = s.lower_bound(3);   // first element >= 3
    auto ub = s.upper_bound(3);   // first element > 3
    cout << *lb << " " << *ub << endl;

    // ════════════════════════════════════
    //  unordered_set<T>  — hash set, O(1) average
    // ════════════════════════════════════
    unordered_set<string> seen;
    seen.insert("apple");
    seen.insert("banana");
    seen.insert("apple");   // duplicate ignored

    cout << seen.count("apple") << endl;   // 1
    cout << seen.size()         << endl;   // 2

    // Remove duplicates from a vector using unordered_set
    vector<int> vec = {1, 2, 2, 3, 3, 3, 4};
    unordered_set<int> unique(vec.begin(), vec.end());
    vector<int> deduped(unique.begin(), unique.end());
    // Note: order not guaranteed with unordered_set

    // ════════════════════════════════════
    //  queue<T>  — FIFO
    // ════════════════════════════════════
    queue<string> q;
    q.push("first");
    q.push("second");
    q.push("third");

    cout << q.front() << endl;   // first (doesn't remove)
    cout << q.back()  << endl;   // third
    q.pop();                     // removes front
    cout << q.front() << endl;   // second
    cout << q.size()  << endl;   // 2

    // ════════════════════════════════════
    //  stack<T>  — LIFO
    // ════════════════════════════════════
    stack<int> stk;
    stk.push(1);
    stk.push(2);
    stk.push(3);

    cout << stk.top() << endl;   // 3 (doesn't remove)
    stk.pop();
    cout << stk.top() << endl;   // 2
    cout << stk.size()<< endl;   // 2

    // ════════════════════════════════════
    //  priority_queue<T>  — max-heap by default
    // ════════════════════════════════════
    priority_queue<int> maxHeap;
    for (int n : {5, 2, 8, 1, 9})
        maxHeap.push(n);

    while (!maxHeap.empty()) {
        cout << maxHeap.top() << " ";   // 9 8 5 2 1
        maxHeap.pop();
    }
    cout << endl;

    // Min-heap
    priority_queue<int, vector<int>, greater<int>> minHeap;
    for (int n : {5, 2, 8, 1, 9})
        minHeap.push(n);

    while (!minHeap.empty()) {
        cout << minHeap.top() << " ";   // 1 2 5 8 9
        minHeap.pop();
    }
    cout << endl;

    // ════════════════════════════════════
    //  deque<T>  — double-ended queue
    // ════════════════════════════════════
    deque<int> dq = {2, 3, 4};
    dq.push_front(1);   // {1, 2, 3, 4}
    dq.push_back(5);    // {1, 2, 3, 4, 5}
    dq.pop_front();     // {2, 3, 4, 5}
    dq.pop_back();      // {2, 3, 4}

    for (int n : dq) cout << n << " ";
    cout << endl;

    // ════════════════════════════════════
    //  list<T>  — doubly linked list
    // ════════════════════════════════════
    list<int> lst = {1, 2, 3, 4, 5};
    lst.push_front(0);    // {0, 1, 2, 3, 4, 5}
    lst.push_back(6);     // {0, 1, 2, 3, 4, 5, 6}
    lst.pop_front();      // {1, 2, 3, 4, 5, 6}

    auto lit = find(lst.begin(), lst.end(), 3);
    lst.insert(lit, 99);   // insert 99 before 3
    lst.erase(lit);        // erase 3

    lst.sort();            // sort in place (list has own sort)
    lst.reverse();
    lst.unique();          // remove consecutive duplicates

    for (int n : lst) cout << n << " ";
    cout << endl;

    return 0;
}
